src/main/java/com/spikes2212/control/RioPIDLoop.java

```diff
@@ -0,0 +1,180 @@
+package com.spikes2212.control;
+
+import edu.wpi.first.wpilibj.Notifier;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.controller.PIDController;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * A PID loop that runs on a separated thread.
+ *
+ * @author Yuval Levy
+ */
+
+public class RioPIDLoop implements PIDLoop {
+
+    /**
+     * An enum that represents the frequency the PID loop runs in.
+     */
+    private enum Frequency {
+        LOW(0.05),
+        DEFAULT(0.02),
+        HIGH(0.01);
+
+        private double period;
+
+        Frequency(double period) {
+            this.period = period;
+        }
+    }
+
+    /**
+     * The notifier creates a separated thread for the PID loop.
+     */
+    private Notifier notifier;
+
+    /**
+     * The PIDController calculates the wanted speed to reach target.
+     */
+    private PIDController controller;
+    /**
+     * The proportional component of the loop.
+     */
+    private Supplier<Double> kp;
+
+    /**
+     * The integral component of the loop.
+     */
+    private Supplier<Double> ki;
+
+    /**
+     * The derivative component of the loop.
+     */
+    private Supplier<Double> kd;
+
+    /**
+     * The acceptable distance from target.
+     */
+    private Supplier<Double> tolerance;
+
+    /**
+     * The setpoint for the loop.
+     */
+    private Supplier<Double> setpoint;
+
+    /**
+     * The time required to stay on target.
+     */
+    private Supplier<Double> waitTime;
+
+    /**
+     * The frequency the PID loop runs in.
+     */
+    private Frequency frequency;
+
+    /**
+     * Returns my current location.
+     */
+    private Supplier<Double> PIDSource;
+
+    /**
+     * The calculation returned from the pid.
+     */
+    private Supplier<Double> PIDCalculations;
```

Comment:
replace with a `Consumer<Double>` named `output`, this consumer should receive the result of the PID calculations done in the notifier thread